src/flows/prospecting/publisher_features.py

```diff
@@ -0,0 +1,77 @@
+import os
+import pandas as pd
+from pathlib import Path
+
+import prefect
+from prefect import Flow, task
+from prefect.executors import LocalDaskExecutor
+
+from api_clients.pocket_snowflake_query import PocketSnowflakeQuery, OutputType
+from common_tasks.load_data import dataframe_to_feature_group
+from utils import config
+from utils.flow import get_flow_name, get_interval_schedule
+
+ONE_DAYS_MINS = 24*60  # using this to schedule flow update interval
+VERSION = 1
+
+FEATURE_GROUP_NAME = f"{config.ENVIRONMENT}-PublisherFeatureGroup-v{VERSION}"
+
+FLOW_NAME = get_flow_name(__file__)
+
+@task()
+def transform_publisher_features(input_df: pd.DataFrame) -> pd.DataFrame:
+
+    required_columns = {"TOP_DOMAIN_NAME", "LOG_TOTAL_SAVES", "TOTAL_SAVE_COUNT", "UPDATED_AT"}
+    if len(required_columns.difference(set(input_df.columns))) > 0:
+        missing = required_columns.difference(set(input_df.columns))
+        raise KeyError(f"input dataframe is missing columns: {missing}")
+
+    # TOP_DOMAIN_NAME is required key for feature group
+    input_df = input_df.dropna(subset=["TOP_DOMAIN_NAME"])
+    # reformat update column
+    input_df["UPDATED_AT"] = input_df.UPDATED_AT.apply(lambda x: x.strftime("%Y-%m-%dT%H:%M:%SZ"))
+    # remove extra column from query
+    input_df = input_df.drop(columns="APPROVED_SOURCE")
+
+    # deal with rows for sources without approved items
+    approved_cols = [c for c in input_df.columns if c.startswith("NUM_APPROVED")]
+    for c in approved_cols:
+        input_df[c] = input_df[c].fillna(0).astype(int)
+
+    return input_df
+
+
+with Flow(FLOW_NAME, schedule=get_interval_schedule(minutes=ONE_DAYS_MINS), executor=LocalDaskExecutor()) as flow:
+
+    logger = prefect.context.get("logger")
+    logger.info(f'Querying snowflake for publisher data')
+
+    # Read a query from the sql/ directory located next to this flow file.
+    sql_path = os.path.join(os.path.dirname(__file__), "sql/publisher_features.sql")
+    sql_query = Path(sql_path).read_text()
+
+    # create query parameters
+    query_params = {"MAX_APPROVED_AGE": 360,
+                    "MIN_WORD_COUNT": 900}
+
+    # retrieve data from snowflake
+    snowflake_result = PocketSnowflakeQuery().run(
+        query=sql_query,
+        data=query_params,
+        output_type=OutputType.DATA_FRAME,
+        database=config.SNOWFLAKE_ANALYTICS_DATABASE,
+        schema=config.SNOWFLAKE_ANALYTICS_DBT_SCHEMA,
+    )
+    print(f"snowflake query returned {len(snowflake_result)} rows")
+
+    # transform data
+    publisher_df = transform_publisher_features.run(snowflake_result)
+
+    # load data
+    result = dataframe_to_feature_group(
+        dataframe=publisher_df,
+        feature_group_name=FEATURE_GROUP_NAME
+    )
```

Comment:
I would suggest writing the flow like this

```suggestion
@task()
def get_snowflake_data():
    logger = prefect.context.get("logger")
    logger.info(f"Querying snowflake for publisher data")

    # Read a query from the sql/ directory located next to this flow file.
    sql_path = os.path.join(os.path.dirname(__file__), "sql/publisher_features.sql")
    sql_query = Path(sql_path).read_text()

    # create query parameters
    query_params = {"MAX_APPROVED_AGE": 360, "MIN_WORD_COUNT": 900}

    # retrieve data from snowflake
    snowflake_result = PocketSnowflakeQuery().run(
        query=sql_query,
        data=query_params,
        output_type=OutputType.DATA_FRAME,
        database=config.SNOWFLAKE_ANALYTICS_DATABASE,
        schema=config.SNOWFLAKE_ANALYTICS_DBT_SCHEMA,
    )
    logger.info(f"snowflake query returned {len(snowflake_result)} rows")
    return snowflake_result


@task()
def transform_publisher_features(input_df: pd.DataFrame) -> pd.DataFrame:

    required_columns = {
        "TOP_DOMAIN_NAME",
        "LOG_TOTAL_SAVES",
        "TOTAL_SAVE_COUNT",
        "UPDATED_AT",
    }
    if len(required_columns.difference(set(input_df.columns))) > 0:
        missing = required_columns.difference(set(input_df.columns))
        raise KeyError(f"input dataframe is missing columns: {missing}")

    # TOP_DOMAIN_NAME is required key for feature group
    input_df = input_df.dropna(subset=["TOP_DOMAIN_NAME"])
    # reformat update column
    input_df["UPDATED_AT"] = input_df.UPDATED_AT.apply(
        lambda x: x.strftime("%Y-%m-%dT%H:%M:%SZ")
    )
    # remove extra column from query
    input_df = input_df.drop(columns="APPROVED_SOURCE")

    # deal with rows for sources without approved items
    approved_cols = [c for c in input_df.columns if c.startswith("NUM_APPROVED")]
    for c in approved_cols:
        input_df[c] = input_df[c].fillna(0).astype(int)

    return input_df


with Flow(
    FLOW_NAME,
    schedule=get_interval_schedule(minutes=ONE_DAYS_MINS),
    executor=LocalDaskExecutor(),
) as flow:

    sf_data_results = get_snowflake_data()

    # transform data
    publisher_df = transform_publisher_features(sf_data_results)

    # load data
    result = dataframe_to_feature_group(
        dataframe=publisher_df, feature_group_name=FEATURE_GROUP_NAME
    )
```

Reply:
this create a flow that looks like this when you execute with `flow.visualize()`:

```python
if __name__ == "__main__":
    # flow.run()
    flow.visualize()
```

![image](https://user-images.githubusercontent.com/11067272/224812924-ac457d55-116d-4459-b534-8b14416eac21.png)

Reply:
prefect will look at your tasks added to the Flow object and execute the run methods...there is no need to call run directly.  Best practice is to create tasks and then stitch them together in the Flow object.

Reply:
this is super helpful, i am still learning basic patterns let alone best practices.

Reply:
sure no worries....Prefect v2 will make it much easier to just write Python in the flow object.  V1 gets wonky when 'untasked' execution elemets are added.

Reply:
i also don't know how to do visualization or make use of the UI, so would be curious about that